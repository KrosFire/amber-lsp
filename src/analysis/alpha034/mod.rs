use tower_lsp::lsp_types::Url;

use crate::{backend::Backend, stdlib::resolve};

pub mod exp;
pub mod global;
pub mod stmnts;

#[tracing::instrument(skip_all)]
pub async fn map_import_path(uri: &Url, path: &str, backend: &Backend) -> Url {
    if path.starts_with("std/") || path == "std" {
        match backend.amber_version {
            _ => {
                if let Some(url) = resolve(backend, "main".to_string()).await {
                    return url;
                }
            } // AmberVersion::Alpha034 if path == "std" => {
              //     if let Some(url) = resolve(backend, "std/main".to_string()).await {
              //         return url;
              //     }
              // }
              // _ => {
              //     if let Some(url) = resolve(backend, path.to_string()).await {
              //         return url;
              //     }
              // }
        }
    }

    let path = uri.to_file_path().unwrap().parent().unwrap().join(path);

    Url::from_file_path(path).unwrap()
}
