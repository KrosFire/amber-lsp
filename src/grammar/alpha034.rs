/// This module contains the grammar for the Amber language.
/// 
/// Precedence levels (Lower on the list means higher precedence):
/// 1. Closure
/// 2. Assignment
/// 3. Range
/// 4. Or
/// 5. And
/// 6. Comparison
/// 7. Additive
/// 8. Multiplicative
/// 9. Cast
/// 10. Unary
/// 11. Call
#[rust_sitter::grammar("Amber_v0_3_4_alpha")]
pub mod grammar {
    #[derive(PartialEq, Debug)]
    pub enum Expression {
        Number(
            #[rust_sitter::leaf(
                pattern = r"\-?\d+(\.\d+)?",
                transform = |v| v.parse::<f32>().unwrap()
            )]
            f32
        ),
        Boolean(Boolean),
        Text(
            #[rust_sitter::leaf(text = "\"")] (),
            Vec<InterpolatedText>,
            #[rust_sitter::leaf(text = "\"")] ()
        ),
        #[rust_sitter::prec(1)]
        Parentheses(
            #[rust_sitter::leaf(text = "(")] (),
            Box<Expression>,
            #[rust_sitter::leaf(text = ")")] ()
        ),
        VariableGet(
            #[rust_sitter::leaf(pattern = r"[a-zA-Z_][a-zA-Z0-9_]*", transform = |v| v.to_string())]
            String
        ),
        #[rust_sitter::prec_left(7)]
        Add(
            Box<Expression>,
            #[rust_sitter::leaf(text = "+")] (),
            Box<Expression>
        ),
        #[rust_sitter::prec_left(7)]
        Subtract(
            Box<Expression>,
            #[rust_sitter::leaf(text = "-")] (),
            Box<Expression>
        ),
        #[rust_sitter::prec_left(8)]
        Multiply(
            Box<Expression>,
            #[rust_sitter::leaf(text = "*")] (),
            Box<Expression>
        ),
        #[rust_sitter::prec_left(8)]
        Divide(
            Box<Expression>,
            #[rust_sitter::leaf(text = "/")] (),
            Box<Expression>
        ),
        #[rust_sitter::prec_left(8)]
        Modulo(
            Box<Expression>,
            #[rust_sitter::leaf(text = "%")] (),
            Box<Expression>
        ),
        #[rust_sitter::prec_right(10)]
        Neg(
            #[rust_sitter::leaf(text = "-")] (),
            Box<Expression>
        ),
        #[rust_sitter::prec_left(5)]
        And(
            Box<Expression>,
            #[rust_sitter::leaf(text = "and")] (),
            Box<Expression>
        ),
        #[rust_sitter::prec_left(4)]
        Or(
            Box<Expression>,
            #[rust_sitter::leaf(text = "or")] (),
            Box<Expression>
        ),
        #[rust_sitter::prec_left(6)]
        Gt(
            Box<Expression>,
            #[rust_sitter::leaf(text = ">")] (),
            Box<Expression>
        ),
        #[rust_sitter::prec_left(6)]
        Ge(
            Box<Expression>,
            #[rust_sitter::leaf(text = ">=")] (),
            Box<Expression>
        ),
        #[rust_sitter::prec_left(6)]
        Lt(
            Box<Expression>,
            #[rust_sitter::leaf(text = "<")] (),
            Box<Expression>
        ),
        #[rust_sitter::prec_left(6)]
        Le(
            Box<Expression>,
            #[rust_sitter::leaf(text = "<=")] (),
            Box<Expression>
        ),
        #[rust_sitter::prec_left(6)]
        Eq(
            Box<Expression>,
            #[rust_sitter::leaf(text = "==")] (),
            Box<Expression>
        ),
        #[rust_sitter::prec_left(6)]
        Neq(
            Box<Expression>,
            #[rust_sitter::leaf(text = "!=")] (),
            Box<Expression>
        ),
        #[rust_sitter::prec_right(10)]
        Not(
            #[rust_sitter::leaf(text = "not")] (),
            Box<Expression>
        ),
        #[rust_sitter::prec_right(1)]
        Ternary(
            Box<Expression>,
            #[rust_sitter::leaf(text = "?")] (),
            Box<Expression>,
            #[rust_sitter::leaf(text = ":")] (),
            Box<Expression>
        ),
        #[rust_sitter::prec(11)]
        FunctionInvocation(
            #[rust_sitter::leaf(pattern = r"[a-zA-Z_][a-zA-Z0-9_]*", transform = |v| v.to_string())]
            String,
            #[rust_sitter::leaf(text = "(")] (),
            #[rust_sitter::delimited(
                #[rust_sitter::leaf(text = ",")]
                ()
            )]
            Vec<Expression>,
            #[rust_sitter::leaf(text = ")")] ()
        ),
        Command(
            #[rust_sitter::leaf(text = "$")] (),
            Vec<InterpolatedCommand>,
            #[rust_sitter::leaf(text = "$")] ()
        ),
        Array(
            #[rust_sitter::leaf(text = "[")] (),
            #[rust_sitter::delimited(
                #[rust_sitter::leaf(text = ",")]
                ()
            )]
            Vec<Expression>,
            #[rust_sitter::leaf(text = "]")] ()
        ),
        #[rust_sitter::prec_left(3)]
        Range(
            Box<Expression>,
            #[rust_sitter::leaf(text = "..")] (),
            #[rust_sitter::leaf(text = "=")] Option<()>,
            Box<Expression>,
        ),
        Null(
            #[rust_sitter::leaf(text = "null")] (),
        ),
        #[rust_sitter::prec_left(9)]
        Cast(
            Box<Expression>,
            #[rust_sitter::leaf(text = "as")] (),
            #[rust_sitter::leaf(pattern = r"[a-zA-Z_][a-zA-Z0-9_]*", transform = |v| v.to_string())]
            String,
        ),
        Status(
            #[rust_sitter::leaf(text = "status")] (),
        ),
        #[rust_sitter::prec_right(8)]
        Nameof(
            #[rust_sitter::leaf(text = "nameof")] (),
            Box<Expression>,
        ),
        #[rust_sitter::prec_left(8)]
        Is(
            Box<Expression>,
            #[rust_sitter::leaf(text = "is")] (),            
            #[rust_sitter::leaf(pattern = r"[a-zA-Z_][a-zA-Z0-9_]*", transform = |v| v.to_string())]
            String,
        )
    }

    #[derive(PartialEq, Debug)]
    pub enum InterpolatedCommand {
        #[rust_sitter::prec(3)]
        Escape(
            #[rust_sitter::leaf(text = "\\")] (),
            #[rust_sitter::leaf(pattern = ".", transform = |v| v.to_string())]
            String,
        ),
        CommandOption(
            #[rust_sitter::leaf(pattern = "-{1,2}")] (),
            #[rust_sitter::leaf(pattern = "[a-zA-Z][A-Za-z0-9-_]*", transform = |v| v.to_string())]
            String
        ),
        Expression(
            #[rust_sitter::leaf(text = "{")] (),
            Box<Expression>,
            #[rust_sitter::leaf(text = "}")] (),
        ),
        Text(
            #[rust_sitter::leaf(pattern = r#"[^\\${-]+"#, transform = |v| v.to_string())]
            String,
        ),
    }

    #[derive(PartialEq, Debug)]
    pub enum InterpolatedText {
        #[rust_sitter::prec(3)]
        Escape(
            #[rust_sitter::leaf(text = "\\")] (),
            #[rust_sitter::leaf(pattern = ".", transform = |v| v.to_string())]
            String,
        ),
        Expression(
            #[rust_sitter::leaf(text = "{")] (),
            Box<Expression>,
            #[rust_sitter::leaf(text = "}")] (),
        ),
        Text(
            #[rust_sitter::leaf(pattern = r#"[^\\"{]+"#, transform = |v| v.to_string())]
            String,
        ),
    }

    #[derive(PartialEq, Debug)]
    pub enum Boolean {
        #[rust_sitter::leaf(text = "true")]
        True,
        #[rust_sitter::leaf(text = "false")]
        False,
    }

    #[rust_sitter::extra]
    struct Whitespace {
        #[rust_sitter::leaf(pattern = r"\s")]
        _whitespace: (),
    }

    #[rust_sitter::language]
    #[derive(PartialEq, Debug)]
    pub enum Statement {
        Expression(Box<Expression>),
    }
}
