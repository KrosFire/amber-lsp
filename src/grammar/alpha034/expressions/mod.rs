use chumsky::prelude::*;

use super::{Expression, Statement};

mod and;
mod ternary;
mod or;
mod atom;
mod cast;
mod comparison;
mod is;
mod product;
mod range;
mod sum;
mod unary;

pub fn parse_expr(stmnts: Recursive<char, Statement, Simple<char>>) -> impl Parser<char, Expression, Error = Simple<char>> + '_ {
    recursive(|expr: Recursive<char, Expression, Simple<char>>| {
        ternary::ternary_parser(stmnts, expr)
    })
}