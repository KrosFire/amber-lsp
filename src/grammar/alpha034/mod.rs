use chumsky::{error::Simple, Parser, Stream};
use heraclitus_compiler::prelude::*;
use lexer::{get_rules, Token};

pub mod expressions;
pub mod global;
pub mod lexer;
pub mod parser;
pub mod statements;

#[derive(PartialEq, Debug, Clone)]
pub enum InterpolatedText {
    Escape(Spanned<String>),
    Expression(Box<Spanned<Expression>>),
    Text(Spanned<String>),
}

#[derive(PartialEq, Debug, Clone)]
pub enum Block {
    Block(Vec<Spanned<Statement>>),
}

#[derive(PartialEq, Debug, Clone)]
pub enum FailureHandler {
    Propagate,
    Handle(Vec<Spanned<Statement>>),
}

#[derive(PartialEq, Debug, Clone)]
pub enum InterpolatedCommand {
    Escape(String),
    CommandOption(String),
    Expression(Box<Spanned<Expression>>),
    Text(String),
}

#[derive(PartialEq, Debug, Clone)]
pub enum Expression {
    Number(Spanned<f32>),
    Boolean(Spanned<bool>),
    Text(Vec<Spanned<InterpolatedText>>),
    Parentheses(Box<Spanned<Expression>>),
    Var(Spanned<String>),
    Add(Box<Spanned<Expression>>, Box<Spanned<Expression>>),
    Subtract(Box<Spanned<Expression>>, Box<Spanned<Expression>>),
    Multiply(Box<Spanned<Expression>>, Box<Spanned<Expression>>),
    Divide(Box<Spanned<Expression>>, Box<Spanned<Expression>>),
    Modulo(Box<Spanned<Expression>>, Box<Spanned<Expression>>),
    Neg(Box<Spanned<Expression>>),
    And(Box<Spanned<Expression>>, Box<Spanned<Expression>>),
    Or(Box<Spanned<Expression>>, Box<Spanned<Expression>>),
    Gt(Box<Spanned<Expression>>, Box<Spanned<Expression>>),
    Ge(Box<Spanned<Expression>>, Box<Spanned<Expression>>),
    Lt(Box<Spanned<Expression>>, Box<Spanned<Expression>>),
    Le(Box<Spanned<Expression>>, Box<Spanned<Expression>>),
    Eq(Box<Spanned<Expression>>, Box<Spanned<Expression>>),
    Neq(Box<Spanned<Expression>>, Box<Spanned<Expression>>),
    Not(Box<Spanned<Expression>>),
    Ternary(
        Box<Spanned<Expression>>,
        Box<Spanned<Expression>>,
        Box<Spanned<Expression>>,
    ),
    FunctionInvocation(
        Spanned<String>,
        Vec<Spanned<Expression>>,
        Option<Spanned<FailureHandler>>,
    ),
    Command(
        Vec<Spanned<InterpolatedCommand>>,
        Option<Spanned<FailureHandler>>,
    ),
    Array(Vec<Spanned<Expression>>),
    Range(Box<Spanned<Expression>>, Box<Spanned<Expression>>),
    Null,
    Cast(Box<Spanned<Expression>>, Spanned<String>),
    Status,
    Nameof(Box<Spanned<Expression>>),
    Is(Box<Spanned<Expression>>, Spanned<String>),
}

#[derive(PartialEq, Debug, Clone)]
pub enum ImportContent {
    ImportAll,
    ImportSpecific(Vec<Spanned<String>>),
}

#[derive(PartialEq, Debug, Clone)]
pub enum FunctionArgument {
    Generic(Spanned<String>),
    Typed(Spanned<String>, Spanned<TypeAnnotation>),
}

#[derive(PartialEq, Debug, Clone)]
pub enum TypeAnnotation {
    Type(Spanned<String>),
}

#[derive(PartialEq, Debug, Clone)]
pub enum IfCondition {
    IfCondition(Box<Spanned<Expression>>, Spanned<Block>),
    InlineIfCondition(Box<Spanned<Expression>>, Box<Spanned<Statement>>),
}

#[derive(PartialEq, Debug, Clone)]
pub enum ElseCondition {
    Else(Spanned<Block>),
    InlineElse(Box<Spanned<Statement>>),
}

#[derive(PartialEq, Debug, Clone)]
pub enum IfChainContent {
    IfCondition(Spanned<IfCondition>),
    Else(Spanned<ElseCondition>),
}

#[derive(PartialEq, Debug, Clone)]
pub enum IterLoopVars {
    Single(Spanned<String>),
    WithIndex(Spanned<String>, Spanned<String>),
}

#[derive(PartialEq, Debug, Clone)]
pub enum CommandModifier {
    Unsafe,
    Silent,
}

#[derive(PartialEq, Debug, Clone)]
pub enum Statement {
    Expression(Box<Spanned<Expression>>),
    VariableInit(Spanned<String>, Box<Spanned<Expression>>),
    VariableSet(Spanned<String>, Box<Spanned<Expression>>),
    IfCondition(Spanned<IfCondition>, Option<Spanned<ElseCondition>>),
    IfChain(Vec<Spanned<IfChainContent>>),
    ShorthandAdd(Spanned<String>, Box<Spanned<Expression>>),
    ShorthandSub(Spanned<String>, Box<Spanned<Expression>>),
    ShorthandMul(Spanned<String>, Box<Spanned<Expression>>),
    ShorthandDiv(Spanned<String>, Box<Spanned<Expression>>),
    ShorthandModulo(Spanned<String>, Box<Spanned<Expression>>),
    InfiniteLoop(Spanned<Block>),
    IterLoop(
        Spanned<IterLoopVars>,
        Box<Spanned<Expression>>,
        Spanned<Block>,
    ),
    Break,
    Continue,
    Return(Option<Box<Spanned<Expression>>>),
    Fail(Option<Box<Spanned<Expression>>>),
    Echo(Box<Spanned<Expression>>),
    CommandModifier(Spanned<CommandModifier>),
    Block(Spanned<Block>),
    Comment(String),
}

#[derive(PartialEq, Debug, Clone)]
pub enum GlobalStatement {
    Import(Spanned<ImportContent>, Spanned<String>),
    FunctionDefinition(
        Spanned<String>,
        Vec<Spanned<FunctionArgument>>,
        Option<Spanned<TypeAnnotation>>,
        Vec<Spanned<Statement>>,
    ),
    Main(Vec<Spanned<Statement>>),
    Statement(Spanned<Statement>),
}

pub type Span = std::ops::Range<usize>;
pub type Spanned<T> = (T, Span);

pub type SpannedSemanticToken = Spanned<usize>;

pub struct AmberCompiler {
    lexer: Compiler,
    parser: Box<dyn Parser<Token, Vec<Spanned<GlobalStatement>>, Error = Simple<Token>>>,
}

impl AmberCompiler {
    pub fn new() -> Self {
        let lexer = Compiler::new("Amber", get_rules());

        let parser = global::global_statement_parser();

        AmberCompiler {
            lexer,
            parser: Box::new(parser),
        }
    }

    pub fn tokenize(&mut self, input: &str) -> Vec<Spanned<Token>> {
        self.lexer.load(input);

        // It should never fail
        self.lexer
            .tokenize()
            .unwrap()
            .iter()
            .map(|t| (Token(t.word.clone()), t.start..(t.start + t.word.len())))
            .collect()
    }

    pub fn parse(&mut self, input: &str) -> (Option<Vec<Spanned<GlobalStatement>>>, Vec<Simple<Token>>) {
        let tokens = self.tokenize(input);
        let len = input.chars().count();
        self.parser.parse_recovery(Stream::from_iter(len..len + 1, tokens.into_iter()))
    }
}
