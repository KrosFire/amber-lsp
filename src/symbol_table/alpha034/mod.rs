use tower_lsp::lsp_types::Url;

use crate::{backend::Backend, stdlib::resolve};

pub mod exp;
pub mod global;
pub mod stmnts;

#[tracing::instrument]
pub fn map_import_path(uri: &Url, path: &str, backend: &Backend) -> Url {
    if path == "std" {
        match resolve(backend, "main") {
            Some(path) => {
                return path;
            }
            None => {}
        }
    }

    let path = uri.to_file_path().unwrap().parent().unwrap().join(path);

    Url::from_file_path(path).unwrap()
}
