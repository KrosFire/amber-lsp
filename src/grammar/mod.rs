use alpha034::GlobalStatement;
use chumsky::error::Simple;

pub mod alpha034;

#[derive(Debug, PartialEq, Clone)]
pub enum Grammar {
    Alpha034(Option<Vec<Spanned<GlobalStatement>>>),
}

pub type Span = std::ops::Range<usize>;
pub type Spanned<T> = (T, Span);
pub type SpannedSemanticToken = Spanned<usize>;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Token(pub String);

impl ToString for Token {
    fn to_string(&self) -> String {
        self.0.clone()
    }
}

impl FromIterator<Token> for String {
    fn from_iter<I: IntoIterator<Item = Token>>(iter: I) -> Self {
        iter.into_iter().map(|t| t.0).collect()
    }
}

#[macro_export]
macro_rules! T {
    [$text:expr] => {
        Token($text.to_string())
    };
}

pub struct ParserResponse {
    pub ast: Grammar,
    pub errors: Vec<Simple<String>>,
    pub semantic_tokens: Vec<SpannedSemanticToken>,
}

pub trait LSPAnalysis: Sync + Send {
    fn parse(&self, input: &str) -> ParserResponse;
}

#[derive(PartialEq)]
pub enum JumpDefinitionResult {
    InFile(Span),
    OpenFile(String),
    None,
}
