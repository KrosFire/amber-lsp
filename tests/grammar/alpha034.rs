use chumsky::error::Simple;
use insta::assert_debug_snapshot;

use amber_lsp::grammar::alpha034::{lexer::Token, AmberCompiler, Spanned};

fn parse(
    input: &str,
) -> (
    Option<Vec<Spanned<amber_lsp::grammar::alpha034::GlobalStatement>>>,
    Vec<Simple<Token>>,
) {
    AmberCompiler::new().parse(input)
}

fn parse_unwrap(input: &str) -> Vec<Spanned<amber_lsp::grammar::alpha034::GlobalStatement>> {
    parse(input).0.unwrap()
}

#[test]
fn test_text() {
    assert_debug_snapshot!(parse_unwrap("\"Hello, world!\""));
    assert_debug_snapshot!(parse_unwrap("\"Hello, {name}!\""));
    assert_debug_snapshot!(parse_unwrap("\"Hello, {name}! How are you?\""));
    assert_debug_snapshot!(parse_unwrap(r#""\"text in quotes\" \\""#));
}

#[test]
fn test_variable_get() {
    assert_debug_snapshot!(parse_unwrap("name"));
    assert_debug_snapshot!(parse_unwrap("name1"));
    assert_debug_snapshot!(parse_unwrap("name_1"));
    assert_debug_snapshot!(parse_unwrap("name_1_"));
    assert_debug_snapshot!(parse_unwrap("_var"));
}

#[test]
fn test_number() {
    assert_debug_snapshot!(parse_unwrap("1"));
    assert_debug_snapshot!(parse_unwrap("1.0"));
    assert_debug_snapshot!(parse_unwrap("-1.0"));
    assert_debug_snapshot!(parse_unwrap("-1.24"));
    assert_debug_snapshot!(parse_unwrap("-5"));
    assert_debug_snapshot!(parse_unwrap("001.00004"));
}

#[test]
fn test_whitespace() {
    assert_debug_snapshot!(parse_unwrap(" 0"));
    assert_debug_snapshot!(parse_unwrap("  0"));
    assert_debug_snapshot!(parse_unwrap("  0 "));
}

#[test]
fn test_bool() {
    assert_debug_snapshot!(parse_unwrap("true"));
    assert_debug_snapshot!(parse_unwrap("false"));
}

#[test]
fn test_add() {
    assert_debug_snapshot!(parse_unwrap("1 + 2"));
    assert_debug_snapshot!(parse_unwrap("1 + 2 + 3"));
}

#[test]
fn test_subtract() {
    assert_debug_snapshot!(parse_unwrap("1 - 2"));
    assert_debug_snapshot!(parse_unwrap("1 - 2 - 3"));
}

#[test]
fn test_add_and_subtract() {
    assert_debug_snapshot!(parse_unwrap("1 + 2 - 3"));
    assert_debug_snapshot!(parse_unwrap("1 - 2 + 3"));
}

#[test]
fn test_multiply() {
    assert_debug_snapshot!(parse_unwrap("1 * 2"));
    assert_debug_snapshot!(parse_unwrap("1 * 2 * 3"));
}

#[test]
fn test_divide() {
    assert_debug_snapshot!(parse_unwrap("1 / 2"));
    assert_debug_snapshot!(parse_unwrap("1 / 2 / 3"));
}

#[test]
fn test_multiply_and_divide() {
    assert_debug_snapshot!(parse_unwrap("1 * 2 / 3"));
    assert_debug_snapshot!(parse_unwrap("1 / 2 * 3"));
}

#[test]
fn test_mults_and_adds() {
    assert_debug_snapshot!(parse_unwrap("1 + 2 * 3"));
    assert_debug_snapshot!(parse_unwrap("1 * 2 + 3"));
    assert_debug_snapshot!(parse_unwrap("1 / 2 + 3"));
    assert_debug_snapshot!(parse_unwrap("1 - 2 / 3"));
    assert_debug_snapshot!(parse_unwrap("3 * 2 - --2 * 2"));
}

#[test]
fn test_modulo() {
    assert_debug_snapshot!(parse_unwrap("1 % 2"));
    assert_debug_snapshot!(parse_unwrap("1 % 2 % 3"));
}

#[test]
fn test_neg() {
    assert_debug_snapshot!(parse_unwrap("-(1)"));
    assert_debug_snapshot!(parse_unwrap("-(1 - 2)"));
}

#[test]
fn test_and() {
    assert_debug_snapshot!(parse_unwrap("true and false"));
    assert_debug_snapshot!(parse_unwrap("true and false and true"));
}

#[test]
fn test_or() {
    assert_debug_snapshot!(parse_unwrap("false or false"));
    assert_debug_snapshot!(parse_unwrap("false or false or true"));
}

#[test]
fn test_gt() {
    assert_debug_snapshot!(parse_unwrap("1 > 2"));
    assert_debug_snapshot!(parse_unwrap("1 + 2 > 2 + 1"));
    assert_debug_snapshot!(parse_unwrap("1 + 2 > 2 + 1 > 5"));
}

#[test]
fn test_ge() {
    assert_debug_snapshot!(parse_unwrap("1 >= 2"));
    assert_debug_snapshot!(parse_unwrap("1 + 2 >= 2 + 1"));
    assert_debug_snapshot!(parse_unwrap("1 + 2 >= 2 + 1 >= 5"));
}

#[test]
fn test_lt() {
    assert_debug_snapshot!(parse_unwrap("1 < 2"));
    assert_debug_snapshot!(parse_unwrap("1 + 2 < 2 + 1"));
    assert_debug_snapshot!(parse_unwrap("1 + 2 < 2 + 1 < 5"));
}

#[test]
fn test_le() {
    assert_debug_snapshot!(parse_unwrap("1 <= 2"));
    assert_debug_snapshot!(parse_unwrap("1 + 2 <= 2 + 1"));
    assert_debug_snapshot!(parse_unwrap("1 + 2 <= 2 + 1 <= 5"));
}

#[test]
fn test_eq() {
    assert_debug_snapshot!(parse_unwrap("1 == 2"));
    assert_debug_snapshot!(parse_unwrap("1 + 2 == 2 + 1"));
    assert_debug_snapshot!(parse_unwrap("1 + 2 == 2 + 1 + 5"));
}

#[test]
fn test_neq() {
    assert_debug_snapshot!(parse_unwrap("1 != 2"));
    assert_debug_snapshot!(parse_unwrap("1 + 2 != 2 + 1"));
    assert_debug_snapshot!(parse_unwrap("1 + 2 != 2 + 1 + 5"));
}

#[test]
fn test_not() {
    assert_debug_snapshot!(parse_unwrap("not true"));
    assert_debug_snapshot!(parse_unwrap("not not true"));
}

#[test]
fn test_ternary() {
    assert_debug_snapshot!(parse_unwrap("   true then 1 else 2   "));
    assert_debug_snapshot!(parse_unwrap("true then 1 + 2 else 2 + 1"));
    assert_debug_snapshot!(parse_unwrap("true then 1 + 2 else false then 5 else 6"));
}

#[test]
fn test_command() {
    assert_debug_snapshot!(parse_unwrap(r#"$echo \"Hello, world!\"$"#));
    assert_debug_snapshot!(parse_unwrap("$echo \"Hello, {name}!\"$"));
    assert_debug_snapshot!(parse("$command --arg1 -v$"));
}

#[test]
fn test_array() {
    assert_debug_snapshot!(parse_unwrap("[1, 2, 3]"));
    assert_debug_snapshot!(parse_unwrap("[1]"));
}

#[test]
fn test_null() {
    assert_debug_snapshot!(parse_unwrap("null"));
}

#[test]
fn test_range() {
    assert_debug_snapshot!(parse_unwrap("1..2"));
    assert_debug_snapshot!(parse_unwrap("1..=2"));
    assert_debug_snapshot!(parse_unwrap("1..2..3"));
}

#[test]
fn test_function_invocation() {
    assert_debug_snapshot!(parse_unwrap("func()"));
    assert_debug_snapshot!(parse_unwrap("func(1)"));
    assert_debug_snapshot!(parse_unwrap("func(1, 2)"));
}

#[test]
fn test_cast() {
    assert_debug_snapshot!(parse_unwrap("1 as Num"));
    assert_debug_snapshot!(parse_unwrap("1 as Num as Text"));
}

#[test]
fn test_nameof() {
    assert_debug_snapshot!(parse_unwrap("nameof variable"));
    assert_debug_snapshot!(parse_unwrap("nameof nameof variable"));
}

#[test]
fn test_expr_precedence() {
    assert_debug_snapshot!(parse_unwrap("1 + 2 * 3"));
    assert_debug_snapshot!(parse_unwrap("1 + 2 / 4 / 6"));
    assert_debug_snapshot!(parse_unwrap("2 - 3 - 4"));
    assert_debug_snapshot!(parse_unwrap("2 - (3 - 4)"));
    assert_debug_snapshot!(parse_unwrap("-(2 + 3) * 5"));
    assert_debug_snapshot!(parse_unwrap("(8+2)*(7-3)/2"));
    assert_debug_snapshot!(parse_unwrap("2 / 3 + 1"));
    assert_debug_snapshot!(parse_unwrap("25 / 5 * 3 + 7 - 2 * 4"));
    assert_debug_snapshot!(parse_unwrap(
        "
            2 + 5 > 3 + 4
                then 15 + 10
                else 5 - 4 <= 1/2
                    then 3 * 4
                    else 2"
    ));
    assert_debug_snapshot!(parse_unwrap("true or false and true and true or false"));
    assert_debug_snapshot!(parse_unwrap("true as Bool as Text as Num * 2 / foo()"));
}

#[test]
fn test_comment() {
    assert_debug_snapshot!(parse_unwrap(
        "
        // This is a comment
        1 + 2
    "
    ));
    assert_debug_snapshot!(parse_unwrap("1 + 2 // This is a comment without a newline"));
}

#[test]
fn test_import() {
    assert_debug_snapshot!(parse_unwrap("import * \"path/to/module\""));
    assert_debug_snapshot!(parse_unwrap("import {} \"path/to/module\""));
    assert_debug_snapshot!(parse_unwrap("import { var1 } \"path/to/module\""));
    assert_debug_snapshot!(parse_unwrap("import { var1, var2 } \"path/to/module\""));
}

#[test]
fn test_function_def() {
    assert_debug_snapshot!(parse_unwrap("fun func() {}"));
    assert_debug_snapshot!(parse_unwrap("fun func(a) {}"));
    assert_debug_snapshot!(parse_unwrap("fun func(a : Num) {}"));
    assert_debug_snapshot!(parse_unwrap("fun func(a: Num, b, c: Bool): Num {}"));
    assert_debug_snapshot!(parse_unwrap(
        "
        fun func(a: Num, b: Text, c: Bool): Num {
            echo 10

            return 10
        }
    "
    ));
}

#[test]
fn test_main_block() {
    assert_debug_snapshot!(parse_unwrap(
        "
        main {
            echo 10
        }

        main {
            echo 3;
        }
    "
    ));
}

#[test]
fn test_var_init() {
    assert_debug_snapshot!(parse_unwrap("let a = 10"));
    assert_debug_snapshot!(parse_unwrap("let a = 10 + 2"));
    assert_debug_snapshot!(parse_unwrap("let a = 10 + 2 * 3"));
}

#[test]
fn test_var_set() {
    assert_debug_snapshot!(parse_unwrap("a = 10"));
    assert_debug_snapshot!(parse_unwrap("a = 10 + 2"));
    assert_debug_snapshot!(parse_unwrap("a = 10 + 2 * 3"));
}

#[test]
fn test_if_condition() {
    assert_debug_snapshot!(parse_unwrap("if true {}"));
    assert_debug_snapshot!(parse_unwrap("if true { echo 10 }"));
    assert_debug_snapshot!(parse_unwrap("if true { echo 10 } else {}"));
    assert_debug_snapshot!(parse_unwrap("if true { echo 10 } else { echo 20 }"));
    assert_debug_snapshot!(parse_unwrap(
        "
        if true: echo 10
        else: echo 20
    "
    ));
}

#[test]
fn test_if_chain() {
    assert_debug_snapshot!(parse_unwrap(
        "
        if {
            2 == 3 {
                echo 10
            }
            true: echo 10
            else {
                echo 20
            }
        }
    "
    ));
}

#[test]
fn test_semicolon() {
    assert_debug_snapshot!(parse_unwrap("1;"));
    assert_debug_snapshot!(parse_unwrap("1; 2;"));
    assert_debug_snapshot!(parse_unwrap(
        "
        main {
            echo 10;
            echo 20

            echo 30;

            10 20 30
        }
    "
    ));
}

#[test]
fn test_shorthands() {
    assert_debug_snapshot!(parse_unwrap("a += 10"));
    assert_debug_snapshot!(parse_unwrap("a -= 10"));
    assert_debug_snapshot!(parse_unwrap("a *= 10"));
    assert_debug_snapshot!(parse_unwrap("a /= 10"));
    assert_debug_snapshot!(parse_unwrap("a %= 10"));
}

#[test]
fn test_loops() {
    assert_debug_snapshot!(parse_unwrap("loop {}"));
    assert_debug_snapshot!(parse_unwrap("loop var1 in 1..2 {}"));
    assert_debug_snapshot!(parse_unwrap("loop var1, var2 in 1..2 {}"));
}

#[test]
fn test_keywords() {
    assert_debug_snapshot!(parse_unwrap("break"));
    assert_debug_snapshot!(parse_unwrap("continue"));
    assert_debug_snapshot!(parse_unwrap("fail"));
    assert_debug_snapshot!(parse_unwrap("fail 1"));
    assert_debug_snapshot!(parse_unwrap("echo 1"));
    assert_debug_snapshot!(parse_unwrap("return"));
    assert_debug_snapshot!(parse_unwrap("return 1"));
}

#[test]
fn test_modifiers() {
    assert_debug_snapshot!(parse_unwrap("unsafe"));
    assert_debug_snapshot!(parse_unwrap("silent"));
    assert_debug_snapshot!(parse_unwrap("silent unsafe {}"));
    assert_debug_snapshot!(parse_unwrap("unsafe silent {}"));
    assert_debug_snapshot!(parse_unwrap("unsafe silent $command$"));
}

#[test]
fn test_failure_handlers() {
    assert_debug_snapshot!(parse_unwrap("$$?"));
    assert_debug_snapshot!(parse_unwrap("$$ failed {}"));
    assert_debug_snapshot!(parse_unwrap("foo()?"));
    assert_debug_snapshot!(parse_unwrap("foo() failed {}"));
}

#[test]
fn test_blocks() {
    assert_debug_snapshot!(parse_unwrap(
        "
        {
            echo 10
        }
        {
            {
                echo 20
            }
        }
    "
    ));
}

#[test]
fn test_recovery() {
    // TODO: Add more tests
    assert_debug_snapshot!(parse("fun foo(abc!) {}"));
    assert_debug_snapshot!(parse(
        "
    5 + 5 +;
    echo 10"
    ));
}

#[test]
fn test_lexer() {
    let mut compiler = AmberCompiler::new();

    assert_debug_snapshot!(compiler.tokenize(
        r#"
        let x = "my \"interpolated\" string {name} end";

        $this --should be - tokenized \$$
        "unclosed string

        abcd {let x = 10
    "#
    ));
}
