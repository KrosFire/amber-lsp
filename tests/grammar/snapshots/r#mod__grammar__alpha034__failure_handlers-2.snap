---
source: tests/grammar/alpha034.rs
expression: "parse_unwrap(\"$$ failed {}\")"
---
[
    (
        Statement(
            (
                Expression(
                    (
                        Command(
                            [],
                            Some(
                                (
                                    Handle(
                                        [],
                                    ),
                                    3..12,
                                ),
                            ),
                        ),
                        0..12,
                    ),
                ),
                0..12,
            ),
        ),
        0..12,
    ),
]
